File: lib/ui.ts
import type { SerializedGame, SerializedProposition } from "./schemas";

/**
 * Instructions explaining how the game format works and how to solve it.
 * These instructions are included in the prompt to help the LLM understand
 * the logical constraints and deduction process.
 */
const gameFormatInstructions = `
## How to Solve This Puzzle

This is a logic deduction puzzle. Your goal: identify which ONE person is the killer.

**The Core Logic:**
- If someone can be proven to have been with ANY attribute (technology, place, company, institution, food, or material), they have an ALIBI
- Having an alibi means you are NOT the killer
- The killer is the ONLY person who cannot be proven to have any alibi
- Eliminate everyone with alibis to find the killer

**Reading Propositions:**

DIRECT ALIBIS (person is definitely NOT the killer):
- "X was with Y" → Person X has alibi, so X ≠ killer
- "X was with Y (alibi: not the killer)" → Explicit alibi for X

CONDITIONAL ALIBIS:
- "If X was with Y, then X is not the killer" → IF you can prove X was with Y, then X has alibi

DISJUNCTIONS (OR statements):
- "(A with X) OR (B with Y)" → At least one is true. Cannot directly eliminate anyone unless you can prove one side false
- "(A with material M) OR (B with food F and institution I)" → Similar logic for complex ORs

**Step-by-Step Method:**
1. First pass: Mark all people with direct alibis as NOT the killer
2. Second pass: Look for OR statements where you can deduce which branch must be true
3. Third pass: Apply conditional statements if their conditions are met
4. The remaining person with NO alibis is the killer

**Critical:** Every suspect except ONE will have at least one alibi. Find the person who has ZERO alibis.

Think carefully and work through each proposition systematically before deciding.
`.trim();

/**
 * Renders a single proposition to a human-readable string
 */
export function renderProposition(prop: SerializedProposition): string {
  const propType = prop.prop_type;

  if (propType === "person_and_attribute") {
    return `${prop.person} was with ${prop.value}`;
  }

  if (propType === "person_or_person") {
    return `(${prop.person1} with ${prop.val1}) OR (${prop.person2} with ${prop.val2})`;
  }

  if (propType === "person_attribute_implies_not_killer") {
    return `If ${prop.person} was with ${prop.value}, then ${prop.person} is not the killer`;
  }

  if (propType === "complex_or") {
    return `(${prop.person1} with ${prop.mat1}) OR (${prop.person2} with ${prop.food2} and ${prop.inst2})`;
  }

  if (propType === "direct_elimination") {
    return `${prop.person} was with ${prop.value} (alibi: not the killer)`;
  }

  // Exhaustiveness check
  const _exhaustiveCheck: never = propType;
  return _exhaustiveCheck;
}

/**
 * Converts a game to a prompt string for LLM inference.
 * The output includes all propositions in a lossless, numbered format.
 */
export function gameToPrompt(game: SerializedGame): string {
  const lines: string[] = [];

  lines.push("# Clue Logic Puzzle");
  lines.push("");
  lines.push("Determine who the killer is based on the following information:");
  lines.push("");

  lines.push(gameFormatInstructions);
  lines.push("");

  lines.push("## Suspects:");
  for (const name of game.names) {
    lines.push(`- ${name}`);
  }
  lines.push("");

  lines.push("## Propositions:");
  for (let i = 0; i < game.propositions.length; i++) {
    const prop = game.propositions[i];
    const rendered = renderProposition(prop);
    lines.push(`${i + 1}. ${rendered}`);
  }
  lines.push("");

  lines.push("Based on these propositions, who is the killer?");

  return lines.join("\n");
}

File: python-scripts/ui.py
"""UI rendering functions for Clue game propositions."""

from clue_models import (
    PROP_COMPLEX_OR,
    PROP_DIRECT_ELIMINATION,
    PROP_PERSON_AND_ATTRIBUTE,
    PROP_PERSON_ATTRIBUTE_IMPLIES_NOT_KILLER,
    PROP_PERSON_OR_PERSON,
    ClueGame,
    PropositionData,
)


def render_proposition(prop_data: PropositionData) -> str:
    """Render a proposition to a human-readable string."""
    if prop_data.prop_type == PROP_PERSON_AND_ATTRIBUTE:
        return f"{prop_data.person} was with {prop_data.value}"

    elif prop_data.prop_type == PROP_PERSON_OR_PERSON:
        return (
            f"({prop_data.person1} with {prop_data.val1}) OR "
            f"({prop_data.person2} with {prop_data.val2})"
        )

    elif prop_data.prop_type == PROP_PERSON_ATTRIBUTE_IMPLIES_NOT_KILLER:
        return (
            f"If {prop_data.person} was with {prop_data.value}, "
            f"then {prop_data.person} is not the killer"
        )

    elif prop_data.prop_type == PROP_COMPLEX_OR:
        return (
            f"({prop_data.person1} with {prop_data.mat1}) OR "
            f"({prop_data.person2} with {prop_data.food2} and {prop_data.inst2})"
        )

    elif prop_data.prop_type == PROP_DIRECT_ELIMINATION:
        return f"{prop_data.person} was with {prop_data.value} (alibi: not the killer)"

    return f"Unknown proposition type: {prop_data.prop_type}"


def print_scenario(game: ClueGame) -> None:
    """Print the game scenario."""
    print(f"🎯 Ground truth: {game.killer} is the killer")
    print("🔍 Full scenario:")
    for name, activity in game.ground_truth.items():
        marker = "🔪" if name == game.killer else "✅"
        print(f"  {marker} {name}: {activity.model_dump()}")
    print()


def print_propositions(game: ClueGame, verbose: bool = True) -> None:
    """Print all propositions in the game with their descriptions."""
    if not game.propositions:
        print("No propositions yet.")
        return

    print(f"\n📜 Propositions ({len(game.propositions)} total):")
    for i, (_, prop_data) in enumerate(game.propositions, 1):
        desc = render_proposition(prop_data)
        if verbose:
            print(f"  {i}. {desc}")
        else:
            # Just show the count
            pass

    if not verbose:
        print(f"  ({len(game.propositions)} propositions)")
